packages/eslint-plugin/tests/rules/no-misused-promises.test.ts

```diff
@@ -1275,5 +1731,389 @@ consume(...cbs);
+        },
+      ],
+    },
+    {
+      code: `
+abstract class MyAbstractClass {
+  abstract setThing(): void;
+}
+
+abstract class MyAbstractSubclassExtendsMyAbstractClass extends MyAbstractClass {
+  abstract setThing(): Promise<void>;
+}
+      `,
+      errors: [
+        {
+          line: 7,
+          messageId: 'voidReturnHeritageType',
+          data: { heritageTypeName: 'MyAbstractClass' },
+        },
+      ],
+    },
+    {
+      code: `
+abstract class MyAbstractClass {
+  abstract setThing(): void;
+}
+
+interface MyInterfaceExtendsMyAbstractClass extends MyAbstractClass {
+  setThing(): Promise<void>;
+}
+      `,
+      errors: [
+        {
+          line: 7,
+          messageId: 'voidReturnHeritageType',
+          data: { heritageTypeName: 'MyAbstractClass' },
+        },
+      ],
+    },
+    // #endregion
+    // #region checksVoidReturn.subtypes: Extending an interface
+    {
+      code: `
+interface MyInterface {
+  setThing(): void;
+}
+
+class MyInterfaceSubclass implements MyInterface {
+  async setThing(): Promise<void> {
+    await Promise.resolve();
+  }
+}
+      `,
+      errors: [
+        {
+          line: 7,
+          messageId: 'voidReturnHeritageType',
+          data: { heritageTypeName: 'MyInterface' },
+        },
+      ],
+    },
+    {
+      code: `
+interface MyInterface {
+  setThing(): void;
+}
+
+abstract class MyAbstractClassImplementsMyInterface implements MyInterface {
+  abstract setThing(): Promise<void>;
+}
+      `,
+      errors: [
+        {
+          line: 7,
+          messageId: 'voidReturnHeritageType',
+          data: { heritageTypeName: 'MyInterface' },
+        },
+      ],
+    },
+    {
+      code: `
+interface MyInterface {
+  setThing(): void;
+}
+
+interface MySubInterface extends MyInterface {
+  setThing(): Promise<void>;
+}
+      `,
+      errors: [
+        {
+          line: 7,
+          messageId: 'voidReturnHeritageType',
+          data: { heritageTypeName: 'MyInterface' },
+        },
+      ],
+    },
+    // #endregion
+    // #region checksVoidReturn.subtypes: Extending type aliases
+    {
+      code: `
+type MyTypeIntersection = { setThing(): void } & { thing: number };
+
+class MyClassImplementsMyTypeIntersection implements MyTypeIntersection {
+  thing = 1;
+  async setThing(): Promise<void> {
+    await Promise.resolve();
+  }
+}
+      `,
+      errors: [
+        {
+          line: 6,
+          messageId: 'voidReturnHeritageType',
+          data: { heritageTypeName: 'MyTypeIntersection' },
+        },
+      ],
+    },
+    {
+      code: `
+type MyGenericType<IsAsync extends boolean = true> = IsAsync extends true
+  ? { setThing(): Promise<void> }
+  : { setThing(): void };
+
+interface MyAsyncInterface extends MyGenericType<false> {
+  setThing(): Promise<void>;
+}
+      `,
+      errors: [
+        {
+          line: 7,
+          messageId: 'voidReturnHeritageType',
+          data: { heritageTypeName: '{ setThing(): void; }' },
```

Comment:
I wasn't fully sure about including the `data` in this test case, since it depends on the specific styling produced by `checker.typeToString(heritageType)`, which actually happens to differ from eslint styling (the semicolon is added).

But if it's unlikely to change then it's maybe fine / nice to keep in as a check that this case is formatted nicely (vs e.g. one of my early iterations had `heritageTypeName` outputted as `"__type"` for a case like this).